### PR TITLE
Feat/injective spot with delegation

### PR DESCRIPTION
**Pull request by aarmoa**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Implementation of a new Injective spot connector.
The new connector does not require Gateway (pure Python connector) and has been implemented to trade with delegated accounts (ie: the trading account trades in representation of a granter account, that is the one having the funds).
The current connector implementation does not support market orders (only limit and limit maker). Market orders support will be added in a future version.

In order to use delegated accounts to trade it is necessary to perform a gran onchain. To send the grant transaction please use the script file `hummingbot/connector/exchange/injective_v2/account_delegation_script.py`.

NOTE:
There is a problem with the dYdX library. It has very hard restrictions on the possible versions to use for `web3`, forcing the whole Hummingbot project to use an old version of that library and all other crypto related libraries required by web3 in place. Taking into account that it is critical for any crypto project to use the most updated library versions for security reasons, the restrictions imposed by `dydx-v3-python` are not acceptable. I think the only solutions the Hummingbot project has are:
- Take the code the dYdX connector needs from the library into the connector itself, to not depend on the library
- Remove the dYdX connector until the dYdX team starts maintaining their library properly

In this PR, to solve the library issue and allow the use of newer `web3` versions, I changed `environment.yml` file to point to the branch of a PR sent already to the dYdX library project that solves the issue (https://github.com/dydxprotocol/dydx-v3-python/pull/202). The PR was created in march in all this time the team has not merged it (and they don't show any intentions to merge it in the near future).

**Tests performed by the developer**:
Created unit tests for all new components.
Tested the connector by running pure market making strategies.

**Tips for QA testing**:
Configure a delegate trading account and teste the connector using spot strategies.

PRP: https://snapshot.org/#/hbot-prp.eth/proposal/0xaf8fa07fbd40c0e92fed0b220c922b1f08416e2e8443e3dfd625ed30c89b6416

### CODE REVIEW

**Comment by rapcmia** [comment]
PR update:
- Ran connect now displays `injective_v2`
   image
- Tried to connect `injective_v2` however we are still looking how we can get the `trading subaccount index` and `granter account` with dev and Injective support on discord _(waiting for response)_
   image

---

We also observe this behavior when it is on idle or switched to other window, it would return a `error parsing` on the background resulting on a distorted display

image

https://github.com/hummingbot/hummingbot/assets/73840223/d4dc8d2c-8980-4b46-ad51-51aaf099d88e

Steps to reproduce:
1. Start the client and switch terminal or to another window for couple of minuts
2. Go back to the client again and see `error parsing`
3. Error not available on logs

**Comment by rapcmia** [comment]
Thanks for the info @keithbaum, ill check this with the team 🙇🏼 

---

PR update:
- Able to connect injective_v2 wallet ✅ 
- However for `account_delegation_script` after filling the required values of the script and using `mainnet` im getting the error below
  image
  ```
  (hummingbot) rapcmia@mishka injective_v2 % python account_delegation_script.py
  E0802 14:20:08.254235000 6183514112 hpack_parser.cc:991]               Error parsing 'content-type' metadata: invalid value
  Traceback (most recent call last):
    File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/account_delegation_script.py", line 84, in <module>
      asyncio.get_event_loop().run_until_complete(main())
    File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
      return future.result()
    File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/account_delegation_script.py", line 58, in main
      (sim_res, success) = await client.simulate_tx(sim_tx_raw_bytes)
    File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 341, in simulate_tx
      metadata = await self.load_cookie(type="chain")
    File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 256, in load_cookie
      metadata = await self.renew_cookie(self.chain_cookie, type)
    File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 231, in renew_cookie
      expires_at = cookie.get(f"{self.cookie_type}").get("expires")
  AttributeError: 'NoneType' object has no attribute 'get'
  (hummingbot) rapcmia@mishka injective_v2 %
  ```
- For `testnet` getting the error below:
   ```
  (hummingbot) rapcmia@mishka injective_v2 % python account_delegation_script.py
  <AioRpcError of RPC that terminated with:
	  status = StatusCode.UNKNOWN
	  details = "grantee and granter should be different With gas wanted: '0' and gas used: '10296' "
	  debug_error_string = "UNKNOWN:Error received from peer ipv4:34.138.231.254:443 {grpc_message:"grantee and granter should be different With gas wanted: \'0\' and gas used: \'10296\' ", grpc_status:2, created_time:"2023-08-02T14:29:55.642778+08:00"}"
  >
  (hummingbot) rapcmia@mishka injective_v2 %
   ```

Steps to reproduce:
1. Add the required values for the script
2. Run `python account_delegation_script.py`

**Comment by aarmoa** [comment]
For the second error, the error message is already describing the problem `grantee and granter should be different`.
Please double check that you are using different accounts for granter and grantee.

**Comment by rapcmia** [comment]
Hi @aarmoa thanks for the help. 
I am able to delegate on `testnet` and seems to be successful [635EA19B1DA15](https://testnet.explorer.injective.network/transaction/635EA19B1DA150B7BC8D335CCB1E15CDC6B77E4ED4387FAAD50A7329C12E6258/). Also ok to connect wallet on the client using a portofolio and trading account. 
image


However when we start the PMM strategy we are getting failed to submit order with the error below:

```
2023-08-03 14:25:22,243 - 21826 - hummingbot.connector.exchange.injective_v2.injective_v2_exchange.InjectiveV2Exchange - NETWORK - Error submitting buy LIMIT_MAKER order to Injective_v2_testnet for 2.00000000 ATOM-USDT 8.604.
Traceback (most recent call last):
  File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 334, in create_orders
    result = await self._send_in_transaction(message=delegated_message)
  File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 633, in _send_in_transaction
    simulation_result = await self.query_executor.simulate_tx(tx_byte=signed_transaction_data)
  File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_query_executor.py", line 143, in simulate_tx
    response, success = await self._sdk_client.simulate_tx(tx_byte=tx_byte)
  File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 341, in simulate_tx
    metadata = await self.load_cookie(type="chain")
  File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 256, in load_cookie
    metadata = await self.renew_cookie(self.chain_cookie, type)
  File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 231, in renew_cookie
    expires_at = cookie.get(f"{self.cookie_type}").get("expires")
AttributeError: 'NoneType' object has no attribute 'get'
```
  -  It seems to return the same error as on the previous comment so we tried to delete the `.chain_cookie` as well still no avail
  - Tried different market IDs available getting the same result `{ATOM-USDT,WETH-USDT}`
  
Steps to reproduce:
1. Run account_delegation_script and check the txn to confirm granter and grantee
2. Setup PMM using `injective_v2 testnet` then start. Observe behavior 

[logs_pmm-injective_v2-02.log](https://github.com/hummingbot/hummingbot/files/12247701/logs_pmm-injective_v2-02.log)

---

When running the script for `mainnet`, getting stream removed error
image

Steps to reproduce:
- Add required values for the config for granter and grantee
- Run `python account_delegation_script` then issue occurred

**Review by cardosofede — APPROVED** [review]
Good job King Kode!

**Comment by cardosofede** [comment]
@aarmoa I forgot to mention that you need to add the connector to connector_status

**Comment by cardosofede** [comment]
> Regarding the first issue @rapcmia (AttributeError: 'NoneType' object has no attribute 'get'), check if in the directory that was the working directory when you executed the command there is a file named `.chain_cookie`. If there is please remove it. The file is generated when you connect to one network, but if you connected first to testnet, you can't then connect to mainnet without removing the file (of if you connected first to mainnet you can't connect to testnet then without removing it)

Is possible to add a check in the connect command to remove if it exists?

File: hummingbot/connector/exchange/injective_v2/README.md
Comment (aarmoa) [reply hummingbot/connector/exchange/injective_v2/README.md:16]: Changing this in the next commit

File: hummingbot/connector/exchange/injective_v2/account_delegation_script.py
Comment (aarmoa) [reply hummingbot/connector/exchange/injective_v2/account_delegation_script.py:17]: It would be possible, but would require the script to initialize an instance of the connector. I don't think it would be a good idea for the following reasons:
- It would make the script, intended to be a very straight forward and simple script, more complex.
- permissions delegation between accounts is something delicate. The user should be fully aware of what they are doing from the Injective protocol point of view. They should be explicitly configuring the script with the markets they want to include instead of trusting any translation/transformation logic

File: hummingbot/connector/exchange/injective_v2/account_delegation_script.py
Comment (aarmoa) [reply hummingbot/connector/exchange/injective_v2/account_delegation_script.py:83]: Yes, it is possible. And it is explained in the API documentation https://api.injective.exchange/#authz-msgrevoke. But that is outside the intended scope of this connector (we don't intend to provide automation for Injective accounts management, only the logic required by Hummingbot to operate with Injective)

File: hummingbot/connector/exchange/injective_v2/injective_v2_api_order_book_data_source.py
Comment (aarmoa) [reply hummingbot/connector/exchange/injective_v2/injective_v2_api_order_book_data_source.py:65]: Yes, we are. That is a must in Hummingbot for them to not be removed until they are not in use anymore.

**Comment by aarmoa** [comment]
> > Regarding the first issue @rapcmia (AttributeError: 'NoneType' object has no attribute 'get'), check if in the directory that was the working directory when you executed the command there is a file named `.chain_cookie`. If there is please remove it. The file is generated when you connect to one network, but if you connected first to testnet, you can't then connect to mainnet without removing the file (of if you connected first to mainnet you can't connect to testnet then without removing it)
> 
> Is possible to add a check in the connect command to remove if it exists?

Hi @cardosofede. I will add this as part of the logic in the PR we will be submitting as soon as this is merged. It will be easier that way because in the other branch we have made several changes to the logic that initializes the Injective AsyncClient and the path of the cookie file

**Comment by aarmoa** [comment]
> Hi @aarmoa thanks for the help. I am able to delegate on `testnet` and seems to be successful [635EA19B1DA15](https://testnet.explorer.injective.network/transaction/635EA19B1DA150B7BC8D335CCB1E15CDC6B77E4ED4387FAAD50A7329C12E6258/). Also ok to connect wallet on the client using a portofolio and trading account. image
> 
> However when we start the PMM strategy we are getting failed to submit order with the error below:
> 
> ```
> 2023-08-03 14:25:22,243 - 21826 - hummingbot.connector.exchange.injective_v2.injective_v2_exchange.InjectiveV2Exchange - NETWORK - Error submitting buy LIMIT_MAKER order to Injective_v2_testnet for 2.00000000 ATOM-USDT 8.604.
> Traceback (most recent call last):
>   File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 334, in create_orders
>     result = await self._send_in_transaction(message=delegated_message)
>   File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 633, in _send_in_transaction
>     simulation_result = await self.query_executor.simulate_tx(tx_byte=signed_transaction_data)
>   File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_query_executor.py", line 143, in simulate_tx
>     response, success = await self._sdk_client.simulate_tx(tx_byte=tx_byte)
>   File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 341, in simulate_tx
>     metadata = await self.load_cookie(type="chain")
>   File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 256, in load_cookie
>     metadata = await self.renew_cookie(self.chain_cookie, type)
>   File "/Users/rapcomia/anaconda3/envs/hummingbot/lib/python3.10/site-packages/pyinjective/async_client.py", line 231, in renew_cookie
>     expires_at = cookie.get(f"{self.cookie_type}").get("expires")
> AttributeError: 'NoneType' object has no attribute 'get'
> ```
> 
> * It seems to return the same error as on the previous comment so we tried to delete the `.chain_cookie` as well still no avail
> * Tried different market IDs available getting the same result `{ATOM-USDT,WETH-USDT}`
> 
> Steps to reproduce:
> 
> 1. Run account_delegation_script and check the txn to confirm granter and grantee
> 2. Setup PMM using `injective_v2 testnet` then start. Observe behavior
> 
> [logs_pmm-injective_v2-02.log](https://github.com/hummingbot/hummingbot/files/12247701/logs_pmm-injective_v2-02.log)
> 
> When running the script for `mainnet`, getting stream removed error image
> 
> Steps to reproduce:
> 
> * Add required values for the config for granter and grantee
> * Run `python account_delegation_script` then issue occurred

The only way to solve the issue you mention is to remove the `.chain_cookie` file.


Regarding the issue with the delegation script in mainnet, try changing the following lines:
`NETWORK = Network.mainnet("sentry3")`
and
`client = AsyncClient(NETWORK, insecure=True)`

**Comment by rapcmia** [comment]
PR update:
- Test latest commit `dddf40e15808f3fa0bfd8c6b8`
- Setup account_delegation for testnet ✅ 

---

Ran test on `injective_v2 testnet`
- Connect wallet successfully
- Setup simple PMM for INJ-USDT 
- Orders are create and cancelled ok 
- Check grantee's txns
   image
   Confirm orders are listed on this example: https://testnet.explorer.injective.network/transaction/712B75E3BA7EABF0903773E7E0182B4DE7249E14C5486AC06D87592FBE9800DA/ ✅

However there are instances of errors on fetching status of order and normally fix itself on the next order cycle
image
  - There are instances that when this happens and stopped the client, there were orders left on the exchange:
     - Tried to start the strategy again see if it can recognized but it created a new set of orders again. Then stopped the strategy and was able to the orders created but not the stuck orders on exchange
     - Tried to relaunch the client but still it did not able to recognize the previous orders

Steps to reproduce:
1. Setup `injective_v2 testnet` on PMM
2. When the error happen while creating orders, stop the strategy then start again 
3. Observe that orders are left on the exchange
    - Another good sample to reproduce this issue is by enabling hanging orders 

[6493.zip](https://github.com/hummingbot/hummingbot/files/12257065/6493.zip)

---

Ran test on `injective_v2 mainnet`
- Run a`ccount_delegation_script` and add `{INJ/USDT, ATOM/USDT}`
- Advised by dev to use and change lines 10 and 29 then run the script
   image
   image
   - txn [36171C91C9D88455E28B079A6BCD2DB0245581535D0F634AC4FC41EF612453F5](https://explorer.injective.network/transaction/36171C91C9D88455E28B079A6BCD2DB0245581535D0F634AC4FC41EF612453F5/) ✅
- Setup PMM using INJ-USDT
   image
   - While trying to create order it now return an error of `Stream removed`
   - Tried to clear the .chain_cookie if this helps but no go
   - Tried to check if changing from sentry3 to {sentry1, sentry 0} but no go (
        - accidentally found this error when setting up account_delegation script so we tried to use it
        image
    ```
    2023-08-04 12:23:38,096 - 5028 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (INJ-USDT) Creating 1 bid orders at (Size, Price): ['1.5 INJ, 8.024 USDT']
    2023-08-04 12:23:38,097 - 5028 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (INJ-USDT) Creating 1 ask orders at (Size, Price): ['1.5 INJ, 8.187 USDT']
    2023-08-04 12:23:39,053 - 5028 - pyinjective.composer.Composer - INFO - Loaded market metadata for: 'Mainnet Spot INJ/USDT'
    2023-08-04 12:23:39,054 - 5028 - pyinjective.composer.Composer - INFO - Loaded market metadata for: 'Mainnet Spot INJ/USDT'
    2023-08-04 12:23:40,738 - 5028 - hummingbot.connector.exchange.injective_v2.injective_v2_exchange.InjectiveV2Exchange - NETWORK - Error submitting buy LIMIT_MAKER order to Injective_v2 for 1.500000000000000000 INJ-USDT 8.024.
    Traceback (most recent call last):
      File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 334, in create_orders
        result = await self._send_in_transaction(message=delegated_message)
      File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 633, in _send_in_transaction
        simulation_result = await self.query_executor.simulate_tx(tx_byte=signed_transaction_data)
      File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_query_executor.py", line 145, in simulate_tx
        raise RuntimeError(f"Transaction simulation failure ({response})")
    RuntimeError: Transaction simulation failure (<AioRpcError of RPC that terminated with:
	    status = StatusCode.UNKNOWN
	    details = "Stream removed"
	    debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"Stream removed", grpc_status:2, created_time:"2023-08-04T12:23:40.737475+08:00"}"
    >)
    2023-08-04 12:23:40,739 - 5028 - hummingbot.connector.exchange.injective_v2.injective_v2_exchange.InjectiveV2Exchange - WARNING - Failed to submit buy order to Injective_v2. Check API key and network connection.
    2023-08-04 12:23:40,740 - 5028 - hummingbot.connector.exchange.injective_v2.injective_v2_exchange.InjectiveV2Exchange - NETWORK - Error submitting sell LIMIT_MAKER order to Injective_v2 for 1.500000000000000000 INJ-USDT 8.187.
    Traceback (most recent call last):
      File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 334, in create_orders
        result = await self._send_in_transaction(message=delegated_message)
      File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_data_source.py", line 633, in _send_in_transaction
        simulation_result = await self.query_executor.simulate_tx(tx_byte=signed_transaction_data)
      File "/Users/rapcomia/github/hummingbot/6493/hummingbot/connector/exchange/injective_v2/injective_query_executor.py", line 145, in simulate_tx
        raise RuntimeError(f"Transaction simulation failure ({response})")
    RuntimeError: Transaction simulation failure (<AioRpcError of RPC that terminated with:
	    status = StatusCode.UNKNOWN
	    details = "Stream removed"
	    debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"Stream removed", grpc_status:2, created_time:"2023-08-04T12:23:40.737475+08:00"}"
    >)
    2023-08-04 12:23:40,740 - 5028 - hummingbot.connector.exchange.injective_v2.injective_v2_exchange.InjectiveV2Exchange - WARNING - Failed to submit buy order to Injective_v2. Check API key and network connection.
    ```
       
Steps to reproduce:
1. Run delegation script and confirm if successful
2. Setup simple PMM with `injective_v2 mainnet`
3. Start the strategy and observe the behavior

[logs_pmm-injective_v2-mainnet.log](https://github.com/hummingbot/hummingbot/files/12257231/logs_pmm-injective_v2-mainnet.log)

**Comment by aarmoa** [comment]
- In Injective the orders are identified buy the "order hash". The order hash is generated based on the order's data and a particular element of the subaccount called the `nonce`. The nonce is a number that changes as the subaccount operates. The Injective connector has to estimate locally what is going to be the orders' hash, by estimating the changes to the nonce counter. But if at any point the nonce in the bot is not correctly synchronized with the nonce in the chain, the estimated hashes will be incorrect. At this point the connector will re-synchronize the nonce number, but the orders already created in the exchange will have to be manually cancelled by the user, because the connector does not know the orders' hashes.
This is not an issue in the connector, it is a known limitation.

- Regarding the "stream removed" error, this happens when the node used to operate is overloaded. That can happen when using the public nodes. To avoid this situation the user should be running its own node. The connector handles these disconnections correctly and recovers. It should not be considered an error.

- I could not understand the error you mentioned happened when you tried to make the bot use "sentry0", "sentry1" and "sentry3". This is not configurable in the bot. Could you describe the steps you took to do this configuration?

**Pull request merged**
